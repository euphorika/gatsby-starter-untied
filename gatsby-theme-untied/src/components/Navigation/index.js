import React from 'react'
import PropTypes from 'prop-types'
import { Link } from 'gatsby'
import MenuIcon from './menu-icon.js'

import styles from './styles.module.styl'

class Navigation extends React.Component {
  constructor(props) {
    super(props)

    this.state = {
      navigationVisible: false,
      horizontalAlignment: this.props.horizontalAlignment || 'center',
      verticalAlignment: this.props.verticalAlignment || 'center',
      entries: this.props.entries,
    }
  }

  renderNavigationEntries = entries =>
    entries.map((entry, key) => (
      <li key={key}>
        <Link
          to={entry.node.link}
          {...(!!entry.node.title ? { title: entry.node.title } : {})}
          {...(!!entry.node.target ? { target: entry.node.target } : {})}
          activeClassName={styles.active}
        >
          {entry.node.name}
        </Link>
      </li>
    ))

  calculateHorizontalPosition = horizontalAlignment => {
    switch (horizontalAlignment) {
      case 'right':
        return styles.horizontalAlignRight
      case 'left':
        return styles.horizontalAlignLeft
      case 'center':
      default:
        return styles.horizontalAlignCenter
    }
  }

  calculateVerticalPosition = verticalAlignment => {
    switch (verticalAlignment) {
      case 'top':
        return styles.verticalAlignTop
      case 'bottom':
        return styles.verticalAlignBottom
      case 'center':
      default:
        return styles.verticalAlignCenter
    }
  }

  render() {
    const toggleVisibilityClass = this.state.navigationVisible
      ? `${styles.menu} ${styles.open}`
      : styles.menu
    const horizontalAlignmentClass = this.calculateHorizontalPosition(
      this.state.horizontalAlignment
    )
    const verticalAlignmentClass = this.calculateVerticalPosition(
      this.state.verticalAlignment
    )

    return (
      <div>
        <button
          className={styles.navLink}
          id="nav-link"
          onClick={e => {
            this.setState({ navigationVisible: !this.state.navigationVisible })
          }}
        >
          <MenuIcon
            className={styles.menuIcon}
            fill="#f1f1f1"
            path="M0 0h17.91v1H0zM0 7.2h17.91v1H0zM0 14.39h17.91v1H0z"
            viewBox="0 0 13.37 13.37"
          />
        </button>
        <div
          className={`${toggleVisibilityClass} ${horizontalAlignmentClass} ${verticalAlignmentClass}`}
          id="menu"
        >
          <button
            className={styles.navClose}
            id="nav-close"
            onClick={() => {
              this.setState({ navigationVisible: false })
            }}
          >
            <MenuIcon
              className={styles.menuIcon}
              fill="#f1f1f1"
              path="M13.37.71L12.66 0 6.68 5.98.71 0 0 .71l5.98 5.97L0 12.66l.71.71 5.97-5.98 5.98 5.98.71-.71-5.98-5.98L13.37.71z"
              viewBox="0 0 17.91 15.39"
            />
          </button>
          <nav>
            <ul>{this.renderNavigationEntries(this.state.entries)}</ul>
          </nav>
        </div>
      </div>
    )
  }
}

Navigation.propTypes = {
  navigationVisible: PropTypes.bool,
  horizontalAlignment: PropTypes.oneOf(['right', 'center', 'left']),
  verticalAlignment: PropTypes.oneOf(['top', 'center', 'bottom']),
  entries: PropTypes.array.isRequired,
}

export default Navigation
