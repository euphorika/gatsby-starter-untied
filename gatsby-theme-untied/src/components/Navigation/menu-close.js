import React from 'react'
import PropTypes from 'prop-types'

const MenuClose = ({ color }) => (
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 13.37 13.37">
    <path
      fill={color}
      d="M13.37.71L12.66 0 6.68 5.98.71 0 0 .71l5.98 5.97L0 12.66l.71.71 5.97-5.98 5.98 5.98.71-.71-5.98-5.98L13.37.71z"
    />
  </svg>
)

MenuClose.propTypes = {
  color: PropTypes.string,
}

MenuClose.defaultProps = {
  color: '#f1f1f1',
}

export default MenuClose
