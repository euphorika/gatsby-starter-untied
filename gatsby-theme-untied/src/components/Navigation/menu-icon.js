import React from 'react'
import PropTypes from 'prop-types'
import styles from './styles.module.styl'

const MenuIcon = ({ svg, fill, viewBox, xmlns, xmlnsXlink, path }) => (
  <svg
    className={styles.menuIcon}
    viewBox={viewBox}
    xmlns={xmlns}
    xmlnsXlink={xmlnsXlink}
  >
    <path d={path} fill={fill} />
  </svg>
)

MenuIcon.defaultProps = {
  xmlns: 'http://www.w3.org/2000/svg',
  xmlnsXlink: 'http://www.w3.org/1999/xlink',
}

MenuIcon.propTypes = {
  svg: PropTypes.node,
  fill: PropTypes.string,
  viewBox: PropTypes.string,
  xmlns: PropTypes.string,
  xmlnsXlink: PropTypes.string,
}

export default MenuIcon
