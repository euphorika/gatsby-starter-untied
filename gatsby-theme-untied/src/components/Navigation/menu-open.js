/** @jsx jsx */
import { jsx } from "theme-ui"

const MenuOpen = ({ size = "1em" }) => (
  <svg
    xmlns="http://www.w3.org/2000/svg"
    fill="currentcolor"
    viewBox="0 0 24 24"
    sx={{
      display: "block",
      margin: 0,
      fontSize: 3,
    }}
  >
    <rect width="20" height="1.23" />
    <rect y="8" width="20" height="1.23" />
    <rect y="16" width="20" height="1.23" />
  </svg>
)

export default props => (
  <span
    title="Toggle Menu"
    {...props}
    sx={{
      color: "inherit",
      bg: "transparent",
      m: 0,
      border: 0,
      appearance: "none",
    }}
  >
    <MenuOpen />
  </span>
)
