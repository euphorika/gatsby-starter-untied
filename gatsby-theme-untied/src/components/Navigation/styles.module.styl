@require '../../themes/standard/config/breakpoints'
@require '../../themes/standard/config/colors'
@require '../../themes/standard/components/navigation'

:root
  --menu-bg-color: primary-color
  --menu-link-color: link-color
  --menu-padding: menu-padding

.nav-link, .nav-close
  cursor: pointer
  text-rendering: geometricPrecision
  color: var(--menu-link-color)
  background-color: primary-color
  border: none
  .menu-icon
    background: primary-color
    width: 64px
    height: 70px
    position: absolute
    top: 0
    right: 10px
    padding: 24px
    color: var(--menu-link-color)


.menu
  display: none
  z-index: 100
  background: var(--menu-bg-color)
  position: fixed
  top: 0
  right: 0
  bottom: 0
  left: 0
  padding: var(--menu-padding)

  &.open
    display: flex

  &.horizontal-align-left
    justify-content: flex-start

  &.horizontal-align-right
    justify-content: flex-end

  &.horizontal-align-center
    justify-content: center

  &.vertical-align-top
    align-items: flex-start

  &.vertical-align-center
    align-items: center

  &.vertical-align-bottom
    align-items: flex-end

  a
    color: var(--menu-link-color)
    text-decoration: none

    &:hover,
    &.active
      text-decoration: underline
