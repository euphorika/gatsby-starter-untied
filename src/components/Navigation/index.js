import React from 'react'
import PropTypes from 'prop-types'
import { Link } from 'gatsby'

import styles from './styles.module.styl'

class Navigation extends React.Component {

  constructor(props) {
    super(props)

    this.state = {
      navigationVisible: false,
      entries: this.props.entries
    }
  }

  _renderNavigationEntries(entries) {
    return entries.map((entry, key) => <li key={key}><Link to={entry.to} title={entry.title}>{entry.text}</Link></li>)
  }

  render() {
    return (
      <div>
        <button
          className={styles.navLink}
          id="nav-link"
          onClick={(e) => {this.setState({navigationVisible: !this.state.navigationVisible})}}>
          â˜°
        </button>
        <nav className={this.state.navigationVisible ? styles.menu + ' ' + styles.open : styles.menu} id="menu">
          <ul>
            {this._renderNavigationEntries(this.state.entries)}
          </ul>
        </nav>
      </div>
    )
  }
}

Navigation.propTypes = {
  entries: PropTypes.array.isRequired,
}

export default Navigation
