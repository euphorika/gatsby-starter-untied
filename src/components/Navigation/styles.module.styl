@require '../../themes/standard/config/breakpoints'
@require '../../themes/standard/config/colors'
@require '../../themes/standard/components/navigation'

:root
  --menu-bg-color secondary-color
  --menu-padding menu-padding

.nav-link
  cursor pointer
  text-rendering geometricPrecision

.menu
  display none
  z-index 100
  background var(--menu-bg-color)
  position fixed
  top 0
  right 0
  bottom 0
  left 0
  padding var(--menu-padding)

  .nav-close
    position absolute
    top 10px
    right 10px

  &.open
    display flex

  &.horizontal-align-left
    justify-content flex-start

  &.horizontal-align-right
    justify-content flex-end

  &.horizontal-align-center
    justify-content center

  &.vertical-align-top
    align-items flex-start

  &.vertical-align-center
    align-items center

  &.vertical-align-bottom
    align-items flex-end

  a
    color link-color
