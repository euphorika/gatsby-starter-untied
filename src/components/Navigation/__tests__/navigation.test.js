import React from "React"
import renderer from "react-test-renderer"
import Navigation from "../"

// required array parameter
const navigationEntries = [
  {
    text: 'Menu Entry 1',
    to: '#',
    title: null
  },{
    text: 'Menu Entry 2',
    to: '#',
    title: null
  },{
    text: 'Menu Entry 3',
    to: '#',
    title: null
  },{
    text: 'Menu Entry 4',
    to: '#',
    title: null
  },{
    text: 'Menu Entry 5',
    to: '#',
    title: null
  },{
    text: 'Menu Entry 6',
    to: '#',
    title: null
  }
]

describe("Navigation", () => {
  it("renders correctly", () => {
    const tree = renderer.create(<Navigation entries={navigationEntries} />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
