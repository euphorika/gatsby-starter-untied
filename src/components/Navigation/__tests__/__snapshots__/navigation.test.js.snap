// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation renders correctly 1`] = `
<div>
  <button
    className="navLink"
    id="nav-link"
    onClick={[Function]}
  >
    â˜°
  </button>
  <nav
    className="menu"
    id="menu"
  >
    <ul>
      <li>
        <Link
          title={null}
          to="#"
        >
          Menu Entry 1
        </Link>
      </li>
      <li>
        <Link
          title={null}
          to="#"
        >
          Menu Entry 2
        </Link>
      </li>
      <li>
        <Link
          title={null}
          to="#"
        >
          Menu Entry 3
        </Link>
      </li>
      <li>
        <Link
          title={null}
          to="#"
        >
          Menu Entry 4
        </Link>
      </li>
      <li>
        <Link
          title={null}
          to="#"
        >
          Menu Entry 5
        </Link>
      </li>
      <li>
        <Link
          title={null}
          to="#"
        >
          Menu Entry 6
        </Link>
      </li>
    </ul>
  </nav>
</div>
`;
